presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java

```diff
@@ -478,48 +476,66 @@ private static String getColumnName(Session session, Metadata metadata, TableHan
+                ExpressionOptimizer expressionOptimizer,
+                ConnectorSession connectorSession)
         {
             ImmutableList.Builder<Subfield.PathElement> elements = ImmutableList.builder();
             while (true) {
-                if (expression instanceof SymbolReference) {
-                    return Optional.of(new Subfield(((SymbolReference) expression).getName(), elements.build().reverse()));
+                if (expression instanceof VariableReferenceExpression) {
+                    return Optional.of(new Subfield(((VariableReferenceExpression) expression).getName(), elements.build().reverse()));
                 }
 
-                if (expression instanceof DereferenceExpression) {
-                    DereferenceExpression dereference = (DereferenceExpression) expression;
-                    elements.add(nestedField(dereference.getField().getValue()));
-                    expression = dereference.getBase();
-                }
-                else if (expression instanceof SubscriptExpression) {
-                    SubscriptExpression subscript = (SubscriptExpression) expression;
-                    Expression index = subscript.getIndex();
-                    if (index instanceof Cast) {
-                        index = ((Cast) index).getExpression();
-                    }
-                    if (index instanceof LongLiteral) {
-                        elements.add(new Subfield.LongSubscript(((LongLiteral) index).getValue()));
-                    }
-                    else if (index instanceof StringLiteral) {
-                        elements.add(new Subfield.StringSubscript(((StringLiteral) index).getValue()));
+                if (expression instanceof SpecialFormExpression && ((SpecialFormExpression) expression).getForm() == DEREFERENCE) {
+                    SpecialFormExpression dereference = (SpecialFormExpression) expression;
+                    RowExpression base = dereference.getArguments().get(0);
+                    RowType baseType = (RowType) base.getType();
+
+                    RowExpression indexExpression = expressionOptimizer.optimize(
+                            dereference.getArguments().get(1),
+                            ExpressionOptimizer.Level.OPTIMIZED,
+                            connectorSession);
+
+                    if (indexExpression instanceof ConstantExpression) {
+                        Object index = ((ConstantExpression) indexExpression).getValue();
+                        if (index instanceof Number) {
+                            Optional<String> fieldName = baseType.getFields().get(((Number) index).intValue()).getName();
+                            if (fieldName.isPresent()) {
+                                elements.add(nestedField(fieldName.get()));
+                                expression = base;
+                                continue;
+                            }
+                        }
```

Comment:
This is the main function that was kind of rewritten. The rest of the file's logic was essentially untouched in this refactor. This implementation is *nearly* identical to SubfieldExtractor.toSubfield in presto-hive as the functionality is identical from my understanding of the old code.